{
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "### TF-KOMPAS: Site caller\n",
        "\n",
        "Updated 1/7/20\n",
        "\n",
        "Author: Zachery Mielko\n",
        "\n",
        "Contributors and Collaborators:\n",
        "* Tiffany Ho: optimization review, feedback, and validation\n",
        "* Farica Zhuang: review, feedback, and validation\n",
        "\n",
        "-----\n",
        "\n",
        "The script requires the following dependencies:\n",
        "\n",
        "* pandas\n",
        "* numpy\n",
        "* biopython\n",
        "* pybedtools\n",
        "\n",
        "The script takes the following as input:\n",
        "- Sequence information in the form of:\n",
        "    - A bed file of genomic coordinates with a genome file in FASTA format\n",
        "    - A FASTA file of sequences\n",
        "- An aligned kmer file\n",
        "\n",
        "With the following parameters\n",
        "- Core definition\n",
        "- Escore threshold\n",
        "- If the transcription factor is palindromic or not\n",
        "- A position to center the calls on (kPosition)\n",
        "\n",
        "The script gives the following as **output**:\n",
        "- Bed file of centered sites (Centered_KOMPAS.bed)\n",
        "    - The Bed file is 0 based and of the format start inclusive, end exclusive. So half-open.\n",
        "    - Columns: Chromosome, Start, End, Orientation, Threshold Score\n",
        "        - If given a FASTA file, KOMPAS will use the ID as the first column instead\n",
        "- Log file (optional)\n",
        "- Scoring Table (optional)\n",
        "    - Useful for troubleshooting or validation\n",
        "    - Gives the score for all positions in all sequences as a tsv file. ScoreTable.tsv\n",
        "    - Gives the calls for all sequences in a tsv format. CallTable.tsv\n",
        "\n",
        "By default, the center position will be the midpoint of the core (rounded down), but it can be specifically chosen as an optional parameter.\n",
        "\n",
        "**Core**\n",
        "\n",
        "The core is the range of positions in the model that must be described in order to call a binding site. This is a prior given by the user based on biological context. The longer the core, the more strict the site calling is. It does not affect the threshold score given to a site. \n",
        "\n",
        "**Threshold Score**\n",
        "\n",
        "This is the threshold for when the site would be called. It is a score used for the calling of a site, but not for the scoring of how well a site is bound. \n",
        "\n",
        "If k > core length (if multiple kmers exist that can fully describe the core), then the second score from **maximum** for those kmers is used. This is due to the requirnment of 2 overlapping kmers.\n",
        "\n",
        "If k <= core length (if multiple (or 1) kmers are needed to describe the core fully), then the **minimum** score for those kmers is used because only when threshold is set to that minimum or below would the site be called\n",
        "\n",
        "**Handles both palindromes and non-palindromes.**\n",
        "\n",
        "If the query is for a palindrome, there is an option isPalindome to set to True. All it does is remove - strand calls, since any match on one strand is a match on another. If this is set to false for a palindrome, it should give + and - matches representing both centered positions, or 2 calls per binding site.\n",
        "\n",
        "-------\n",
        "Latest update (1/7/20) - ZM\n",
        "* Input can be either:\n",
        "    - Bed coordinates and genome file\n",
        "    - Fasta file   \n",
        "* Added additional parameter for scoring tables (optional)\n",
        "* Fixed a bug where a site would be reported multiple times with different scores\n",
        "\n",
        "Update (12/5/19) - ZM\n",
        "* Minor bug fix for when k == core length\n",
        "\n",
        "Update (11/27/19) - ZM\n",
        "* Completely redone site calling implementation using numpy arrays.\n",
        "    - Fixes an issue with overlapping cores, especially at low Escore thresholds. \n",
        "    - Now KOMPAS can call overlapping cores in the same manner at any threshold. \n",
        "    - Removes a couple dependencies, though they are packeged with python by default\n",
        "* Fixed a bug with the threshold scoring\n",
        "* Optimized palindrome mode, should run 2x as fast as non-palindrome mode\n",
        "\n",
        "Update (11/25/19) - ZM\n",
        "* Diagnostic tsv file shows all searched peaks, even if no calls were made\n",
        "* Threshold score is now set up so you can use it as a substitute for running at different thresholds. Represents practical threshold given 2 overlapping, not theoretical. \n",
        "\n",
        "\n",
        "Update (11/20/19) - ZM\n",
        "* Fixed bug in coordinates for 0 vs 1 based. KOMPAS is now fully in-sync with half-open, 0 based formats used by bedtools, MACS, and UCSC (backend)\n"
      ],
      "metadata": {}
    },
    {
      "cell_type": "code",
      "source": [
        "################## Parameters ###############\n",
        "wkdir = \"/Users/ZMielko/Desktop/KOMPAS_Paper/gcPBMAnalysis/gcPBM/\"\n",
        "kmerFile = '/Users/ZMielko/Desktop/KOMPAS_Paper/gcPBMAnalysis/TFs_Processed/ETS1/ETS1_8mersAligned.txt'\n",
        "isFasta = True\n",
        "# For using coordinate input, can leave as empty string if using fasta\n",
        "peakFile = ''\n",
        "genomeFile = ''\n",
        "# For using fasta input, can leave as empty string if using bed files\n",
        "fastaFile = '/Users/ZMielko/Desktop/KOMPAS_Paper/gcPBMAnalysis/gcPBM/Ets1.fasta'\n",
        "# kPosition of the core and where to center the call\n",
        "# core is right exclusionary, left inclusive [)\n",
        "core = (10,14) \n",
        "threshold = 0.4\n",
        "isPalindrome = False\n",
        "# Optional settings\n",
        "centerPos = 12 # if set to 'default', will be half position of the core rounded down\n",
        "logFile = True\n",
        "scoreTable = False # for diagnostic/troubleshooting purposes\n",
        "\n",
        "\n",
        "# Calculated parameters and error checking\n",
        "if centerPos == 'default': # Calculate center position if default\n",
        "    centerPos = int((core[0] + core[1])/2)\n",
        "    \n",
        "##### Imports ####\n",
        "\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "from Bio import SeqIO\n",
        "from Bio.Seq import reverse_complement\n",
        "import os\n",
        "from pybedtools import BedTool\n",
        "\n",
        "#### Parsing and Conversion Functions ####\n",
        "def readFasta(file):\n",
        "    '''\n",
        "    Reads a fasta file and turns it into a dataframe with forward and rc sequences\n",
        "    Input: fasta file path\n",
        "    Output: DataFrame with seq_id, fwd sequence, sequence length, reverse complement\n",
        "    '''\n",
        "    entry = []\n",
        "    with open(file, \"r\") as input_handle:\n",
        "        for record in SeqIO.parse(input_handle, \"fasta\"):\n",
        "            entry.append([record.id, str(record.seq).upper(), len(record.seq), str(reverse_complement(record.seq)).upper()])\n",
        "    arr = np.array(entry)\n",
        "    df = pd.DataFrame({'seq_id':arr[:, 0], 'fwd':arr[:,1], 'seq_len':arr[:,2], 'rev_comp':arr[:,3]})\n",
        "    return(df)\n",
        "\n",
        "\n",
        "def parseID(df):\n",
        "    \"\"\"\n",
        "    Takes the concatinated names given in fasta outputs from bedtool's getfasta \n",
        "    and turns them into bed compatible columns\n",
        "    Input: DataFrame from readFasta\n",
        "    Output: DataFrame with Chromosome, Start, End columns\n",
        "    \"\"\"\n",
        "    chrom, start, end = [], [], []\n",
        "    for i in df.seq_id:\n",
        "        cr = i.split(':')\n",
        "        pos = cr[1].split('-')\n",
        "        chrom.append(cr[0])\n",
        "        start.append(int(pos[0]))\n",
        "        end.append(int(pos[1]))\n",
        "    df['Chromosome'] = chrom\n",
        "    df['Start'] = start\n",
        "    df['End'] = end\n",
        "    return(df)\n",
        "\n",
        "def parseIDFASTA(df):\n",
        "    \"\"\"parseIDFASTA takes the fasta ID and converts it into a format comaprable to parseID\n",
        "    Chromosome = ID, Start = 0, End = len(seq)\"\"\"\n",
        "    chrom, start, end = [], [], []\n",
        "    for ID, seq in zip(df.seq_id, df.fwd):\n",
        "        chrom.append(ID)\n",
        "        start.append(0)\n",
        "        end.append(len(seq))\n",
        "    df['Chromosome'] = chrom\n",
        "    df['Start'] = start\n",
        "    df['End'] = end\n",
        "    return(df)\n",
        "\n",
        "def convertToBed(df, isPalindrome):\n",
        "    \"\"\"\n",
        "    Takes a dataframe with the following columns:\n",
        "    Chromosome, Start, centerPlus, scorePlus, centerMinus, scoreMinus\n",
        "    Outputs a dataframe in a bed format \n",
        "    \"\"\"\n",
        "    if isPalindrome == False:\n",
        "        chrom, start, orient,scores = [],[],[],[]\n",
        "        for row in zip(df['Chromosome'],df['Start'],df['centerPlus'],df['scorePlus'],df['centerMinus'],df['scoreMinus']):\n",
        "            if row[2]:\n",
        "                for centerP, score in zip(row[2], row[3]): # + sites\n",
        "                    chrom.append(row[0])\n",
        "                    start.append(row[1] + centerP)\n",
        "                    orient.append('+')\n",
        "                    scores.append(score)\n",
        "            if row[4]:\n",
        "                for centerN, score in zip(row[4],row[5]): # - sites\n",
        "                    chrom.append(row[0])\n",
        "                    start.append(row[1] + centerN)\n",
        "                    orient.append('-')\n",
        "                    scores.append(score)\n",
        "    else:\n",
        "        for row in zip(df['Chromosome'],df['Start'],df['centerPlus'],df['scorePlus']):\n",
        "            if row[2]:\n",
        "                for centerP, score in zip(row[2], row[3]): # + sites\n",
        "                    chrom.append(row[0])\n",
        "                    start.append(row[1] + centerP)\n",
        "                    orient.append('+')\n",
        "                    scores.append(score)\n",
        "    bedDF = pd.DataFrame({'chrom':chrom,'start':start, 'end':start,'orient':orient, 'score':scores})\n",
        "    bedDF['end'] = bedDF['end'] + 1 # exclusive end position\n",
        "    return(bedDF)\n",
        "##### Read in kmer data and process ####\n",
        "kmer = pd.read_csv(kmerFile, sep = '\\t')\n",
        "k = len(kmer['kmer'][0])\n",
        "coreLen = core[1] - core[0]\n",
        "# Find the kPositions required, any would be sufficient to call\n",
        "if k > coreLen: \n",
        "    searchEnd = core[1]\n",
        "    checkK = 0\n",
        "    ReqKpos = set() #\n",
        "    while checkK != core[0]:\n",
        "        checkK = searchEnd - k\n",
        "        if checkK <= core[0]:\n",
        "            ReqKpos.add(checkK)\n",
        "            searchEnd = searchEnd + 1\n",
        "# Or find the group of all kPositions that are needed, all or none\n",
        "else:\n",
        "    searchStart = core[0]\n",
        "    checkK = 0\n",
        "    ReqKpos = set()\n",
        "    while searchStart + k <= core[1]:\n",
        "        ReqKpos.add(searchStart)\n",
        "        searchStart = searchStart + 1\n",
        "# Determine flanks of ReqKPos for threshold score reporting\n",
        "ScoredKpos = ReqKpos.copy()\n",
        "if k >= coreLen:\n",
        "    ScoredKpos.add(min(ReqKpos) - 1)\n",
        "    ScoredKpos.add(max(ReqKpos) + 1)        \n",
        "\n",
        "# Generate dictionary for quick retreaval of relevant kmers\n",
        "thrKmers = kmer[(kmer['Escore'] > threshold) & (kmer['kposition'].isin(ScoredKpos))]\n",
        "kDict = dict(zip(thrKmers['kmer'],zip(thrKmers['kposition'],thrKmers['Escore'])))\n",
        "\n",
        "\n",
        "##### Generate a dataframe from the input sequence information ####\n",
        "if isFasta == True: # if given a FASTA file of sequences    \n",
        "    peakDF = readFasta(fastaFile)\n",
        "    peakDF = parseIDFASTA(peakDF)\n",
        "    del peakDF['seq_id']\n",
        "    peakDF[['seq_len', 'Start','End']] = peakDF[['seq_len', 'Start','End']].astype(int) \n",
        "else: # if given a bed and genome file\n",
        "    peaks = pd.read_csv(peakFile, sep = '\\t',header = None, usecols=[0,1,2])\n",
        "    # find out how long the peaks need to be to use them\n",
        "    if coreLen > k:\n",
        "        minLen = coreLen\n",
        "    else:\n",
        "        minLen = k + 1\n",
        "    peaks = peaks[peaks[2]-peaks[1] > (minLen)].drop_duplicates() # filter for short sequences the caller would have trouble with\n",
        "    # Generate peaks using bedtools\n",
        "    peakBed = BedTool.from_dataframe(peaks)\n",
        "    peakSequence = peakBed.sequence(fi = genomeFile)\n",
        "    peakDF = readFasta(peakSequence.seqfn)\n",
        "    peakDF = parseID(peakDF)\n",
        "    del peakDF['seq_id']\n",
        "    peakDF[['seq_len', 'Start','End']] = peakDF[['seq_len', 'Start','End']].astype(int) \n",
        "\n",
        "\n",
        "##### KOMPAS Scoring and Calling Functions #####   \n",
        "    \n",
        "def kmerMatch(seq):\n",
        "    \"\"\"\n",
        "    Returns matched positions in the sequence and their kpositions\n",
        "    Input: sequence\n",
        "    Output: consecutive positions, kpositions, and scores above threshold\n",
        "    \"\"\"\n",
        "    # Get the kposition and kscore for the peak, save a numpy array\n",
        "    kpos,kscore = [], []\n",
        "    for i in range(len(seq) - k + 1):\n",
        "        window = seq[i:i+k]\n",
        "        if window in kDict:\n",
        "            kpos.append(kDict[window][0])\n",
        "            kscore.append(kDict[window][1])\n",
        "        else:\n",
        "            kpos.append(0)\n",
        "            kscore.append(-0.5)\n",
        "    kpos = np.array(kpos)\n",
        "    kscore = np.array(kscore)\n",
        "    # Get consecutive positions, kpositions, and score via numpy operations\n",
        "    if k >= coreLen:\n",
        "        position = list(filter(lambda x: len(x) != 1,np.split(np.r_[:len(kpos)], np.where(np.diff(kpos) != 1)[0]+1)))\n",
        "        kpos = list(filter(lambda x: len(x) != 1,np.split(kpos, np.where(np.diff(kpos) != 1)[0]+1)))\n",
        "    elif k < coreLen:\n",
        "        reqLen = len(ReqKpos)\n",
        "        position = list(filter(lambda x: len(x) == reqLen,np.split(np.r_[:len(kpos)], np.where(np.diff(kpos) != 1)[0]+1)))\n",
        "        kpos = list(filter(lambda x: len(x) == reqLen,np.split(kpos, np.where(np.diff(kpos) != 1)[0]+1)))\n",
        "    kScore = []\n",
        "    for pos in position:\n",
        "        kScore.append(kscore[pos])\n",
        "    return(zip(position, kpos, kScore))\n",
        "\n",
        "def findCenter(zippedMatch, orient, seqLen):\n",
        "    \"\"\"\n",
        "    Given a zip of match position, kposition, and kscore\n",
        "    Returns the center sites and threshold kscore\n",
        "    \"\"\"\n",
        "    centerSites = []\n",
        "    siteScores = []\n",
        "    for pos, kpos, kScore in zippedMatch:\n",
        "        centerSite = (centerPos - kpos[0]) + pos[0]\n",
        "        if orient == 'rc':\n",
        "            centerSite = (seqLen - centerSite) -1\n",
        "        centerSites.append(centerSite)\n",
        "        if k >= coreLen:\n",
        "            score = threshold\n",
        "            for score1, score2 in zip(kScore, kScore[1:]):\n",
        "                caniScore = sorted([score1, score2])[0]\n",
        "                if caniScore > score:\n",
        "                    score = caniScore\n",
        "            siteScores.append(score)\n",
        "        elif k < coreLen:\n",
        "            siteScores.append(min(kScore))\n",
        "    return(pd.Series([centerSites, siteScores]))\n",
        "             \n",
        "###### Run Calling Functions on Sequence Data ######\n",
        "if isPalindrome == True:\n",
        "    peakDF[[\"centerPlus\",\"scorePlus\"]] = peakDF.apply(lambda peakDF: findCenter(kmerMatch(peakDF['fwd']), 'fwd', len(peakDF['fwd'])), axis = 1) \n",
        "    if scoreTable == True:\n",
        "        peakDF.to_csv(f'{wkdir}CallTable_{threshold}_KOMPAS.tsv', sep = '\\t')\n",
        "    peakDF = peakDF[~peakDF['centerPlus'].isna()]\n",
        "else:\n",
        "    peakDF[[\"centerPlus\",\"scorePlus\"]] = peakDF.apply(lambda peakDF: findCenter(kmerMatch(peakDF['fwd']), 'fwd',len(peakDF['fwd'])), axis = 1) \n",
        "    peakDF[[\"centerMinus\",\"scoreMinus\"]] = peakDF.apply(lambda peakDF: findCenter(kmerMatch(peakDF['rev_comp']), 'rc',len(peakDF['rev_comp'])), axis = 1)\n",
        "    if scoreTable == True:\n",
        "        peakDF.to_csv(f'{wkdir}CallTable_{threshold}_KOMPAS.tsv', sep = '\\t')\n",
        "    peakDF = peakDF[~peakDF['centerPlus'].isna() | ~peakDF['centerMinus'].isna()]\n",
        "    \n",
        "# Convert to Bed format and save\n",
        "finalBed = convertToBed(peakDF, isPalindrome)\n",
        "finalBed = finalBed.sort_values(by='score', ascending = False).drop_duplicates(['chrom', 'start', 'end', 'orient'],keep = 'first')\n",
        "finalBed.to_csv(f'{wkdir}Centered_{threshold}_KOMPAS.bed', sep = '\\t', header = None, index = False)\n",
        "\n",
        "####### Optional Output #######\n",
        "\n",
        "# Score Table Output\n",
        "if scoreTable == True:\n",
        "    def ScoreTableGenerate(seq):\n",
        "        \"\"\"\n",
        "        Returns matched positions in the sequence and their kpositions\n",
        "        Input: sequence \n",
        "        Output: kpositions, and scores above threshold\n",
        "        \"\"\"\n",
        "        # Get the kposition and kscore for the peak, save a numpy array\n",
        "        kpos,kscore = [], []\n",
        "        for i in range(len(seq) - k + 1):\n",
        "            window = seq[i:i+k]\n",
        "            if window in kDict:\n",
        "                kpos.append(kDict[window][0])\n",
        "                kscore.append(kDict[window][1])\n",
        "            else:\n",
        "                kpos.append(0)\n",
        "                kscore.append(-0.5)\n",
        "        kpos = kpos\n",
        "        kscore = kscore\n",
        "        return(kpos, kscore)\n",
        "\n",
        "    fwdScores = peakDF.apply(lambda peakDF: kmerMatchScoreTable(peakDF['fwd']), axis = 1) \n",
        "    rcScores = peakDF.apply(lambda peakDF: kmerMatchScoreTable(peakDF['rev_comp']), axis = 1) \n",
        "    pd.DataFrame({'fwd':fwdScores, 'rc':rcScores}).to_csv(f'{wkdir}ScoreTable_{threshold}.tsv', sep = '\\t')\n",
        "\n",
        "# Log File\n",
        "if logFile == True:\n",
        "    ##################\n",
        "    # Log the output #\n",
        "    ##################\n",
        "    f = open(wkdir + \"/KOMPASLog.txt\", \"a\")\n",
        "    f.write(\"##### Parameters ##### \\n\")\n",
        "    if isFasta == True:\n",
        "        f.write(f\"Fasta file: {fastaFile}\"+ \"\\n\")\n",
        "    else:\n",
        "        f.write(f\"Peak file: {peakFile}\"+ \"\\n\")\n",
        "        f.write(f\"Genome file: {genomeFile}\"+ \"\\n\")\n",
        "    f.write(f\"kmer file: {kmerFile}\"+ \"\\n\")\n",
        "    f.write(f\"Core kPositions: {core}\"+ \"\\n\")\n",
        "    f.write(f\"Center kPositions: {centerPos}\"+ \"\\n\")\n",
        "    f.write(f\"Threshold: {threshold}\"+ \"\\n\")\n",
        "    f.write(\"#Summary# \\n\")\n",
        "    f.write(f\"Final # of called sequences: {len(finalBed)}\"+ \"\\n\")\n",
        "    f.close()\n"
      ],
      "outputs": [],
      "execution_count": null,
      "metadata": {}
    }
  ],
  "metadata": {
    "kernelspec": {
      "name": "python3",
      "language": "python",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python",
      "version": "3.7.3",
      "mimetype": "text/x-python",
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "pygments_lexer": "ipython3",
      "nbconvert_exporter": "python",
      "file_extension": ".py"
    },
    "kernel_info": {
      "name": "python3"
    },
    "nteract": {
      "version": "0.15.0"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}